File: src/main/java/pr/eleks/we_at_her/services/api/impl/AbstractApiServiceImpl.java
package pr.eleks.we_at_her.services.api.impl;

import org.springframework.core.env.Environment;
import org.springframework.stereotype.Service;
import org.springframework.web.util.UriComponentsBuilder;
import pr.eleks.we_at_her.exceptions.PropertyNotFoundException;
import pr.eleks.we_at_her.services.api.ApiService;

import java.util.HashMap;
import java.util.Map;
import java.util.Optional;

@Service
public abstract class AbstractApiServiceImpl implements ApiService {

    protected Map<String, String> defaultValues;
    protected String apiPrefix = "wApis.";
    protected UriComponentsBuilder uriBuilder;
    protected Environment env;


    @Override
    public void prepareParameters(String latitude, String longitude, String lang, String units) throws PropertyNotFoundException {
        // Check parameters
        defaultValues = new HashMap<>();
        defaultValues.put("latitude", latitude.equals("")
                ? Optional.ofNullable(env.getProperty("city.Ternopil.lat"))
                .orElseThrow(() -> new PropertyNotFoundException("city.Ternopil.lat"))
                : latitude);
        defaultValues.put("longitude", longitude.equals("")
                ? Optional.ofNullable(env.getProperty("city.Ternopil.lon"))
                .orElseThrow(() -> new PropertyNotFoundException("city.Ternopil.lon"))
                : longitude);
        defaultValues.put("lang", lang.equals("")
                ? Optional.ofNullable(env.getProperty(apiPrefix + getName() + ".lang"))
                .orElseThrow(() -> new PropertyNotFoundException(apiPrefix + getName() + ".lang"))
                : lang);
        defaultValues.put("units", units.equals("")
                ? Optional.ofNullable(env.getProperty(apiPrefix + getName() + ".units"))
                .orElseThrow(() -> new PropertyNotFoundException(apiPrefix + getName() + ".units"))
                : units);

    }

    @Override
    public void prepareBaseUrl() throws PropertyNotFoundException {
        uriBuilder = UriComponentsBuilder
                .fromUriString(
                        Optional
                                .ofNullable(env.getProperty(apiPrefix + getName() + ".baseUrl"))
                                .orElseThrow(() -> new PropertyNotFoundException(apiPrefix + getName() + ".baseUrl")))
                .pathSegment(
                        Optional
                                .ofNullable(env.getProperty(apiPrefix + getName() + ".request"))
                                .orElseThrow(() -> new PropertyNotFoundException(apiPrefix + getName() + ".request")));
    }

}

File: src/main/kotlin/pr/eleks/we_at_her/services/api/impl/AbstractApiServiceImpl.kt
package pr.eleks.we_at_her.services.api.impl

import org.springframework.core.env.Environment
import org.springframework.stereotype.Service
import org.springframework.web.util.UriBuilder
import org.springframework.web.util.UriComponentsBuilder
import pr.eleks.we_at_her.exceptions.PropertyNotFoundException
import pr.eleks.we_at_her.services.api.ApiService
import java.util.*

@Service
abstract class AbstractApiServiceImpl : ApiService {

    protected lateinit var defaultValues: MutableMap<String, String>
    protected lateinit var uriBuilder: UriComponentsBuilder
    protected lateinit var env: Environment
    protected val apiPrefix: String = "wApis."

    override fun prepareParameters(latitude: String?, longitude: String?, lang: String?, units: String?) {
        // Check parameters
        defaultValues = HashMap()
        defaultValues["latitude"] = if (latitude.isNullOrBlank()) {
            env.getProperty("city.Ternopil.lat")
                    ?: throw PropertyNotFoundException("city.Ternopil.lat")
        } else latitude
        defaultValues["longitude"] = if (longitude.isNullOrBlank()) {
            env.getProperty("city.Ternopil.lon")
                    ?: throw PropertyNotFoundException("city.Ternopil.lon")
        } else longitude
        defaultValues["lang"] = if (lang.isNullOrBlank()) {
            env.getProperty(apiPrefix + getName() + ".lang")
                    ?: throw PropertyNotFoundException(apiPrefix + getName() + ".lang")
        } else lang
        defaultValues["units"] = if (units.isNullOrBlank()) {
            env.getProperty(apiPrefix + getName() + ".units")
                    ?: throw PropertyNotFoundException(apiPrefix + getName() + ".units")
        } else units
    }

    @Throws(PropertyNotFoundException::class)
    override fun prepareBaseUrl() {
        uriBuilder = UriComponentsBuilder
                .fromUriString(
                        env.getProperty(apiPrefix + getName() + ".baseUrl")
                                ?: throw PropertyNotFoundException(apiPrefix + getName() + ".baseUrl"))
                .pathSegment(
                        env.getProperty(apiPrefix + getName() + ".request")
                                ?: throw PropertyNotFoundException(apiPrefix + getName() + ".request"))
    }
}
